libyul/AsmAnalysis.cpp

```diff
@@ -319,6 +319,23 @@ std::vector<YulString> AsmAnalyzer::operator()(FunctionCall const& _funCall)
 				"The underlying opcode will eventually undergo breaking changes, "
 				"and its use is not recommended."
 			);
+		else if (
+			m_evmVersion.supportsTransientStorage() &&
+			(
+				_funCall.functionName.name == "tstore"_yulstring ||
+				_funCall.functionName.name == "tload"_yulstring
+			)
+		)
+			m_errorReporter.warning(
+				2394_error,
+				nativeLocationOf(_funCall.functionName),
+				"Due to a bug in the specification of EIP-1153, transient storage can break the composability of smart contracts. "
+				"Since transient storage is cleared only at the end of the transaction and not at the end of call frames, "
+				"your contract may unintentionally misbehave when invoked multiple times in a complex transaction. "
+				"To avoid this, be sure to clear all transient storage at the end of any call to your contract. "
+				"The use of transient storage for reentrancy guards that are cleared at the end of the call is safe."
```

Comment:
I'm wondering whether we should dial the warning down slightly, to something like:

```suggestion
				"Transient storage can break the composability of smart contracts: "
				"Since transient storage is cleared only at the end of the transaction and not at the end of the outermost call frame to the contract within a transaction, "
				"your contract may unintentionally misbehave when invoked multiple times in a complex transaction. "
				"To avoid this, be sure to clear all transient storage at the end of any call to your contract. "
				"The use of transient storage for reentrancy guards that are cleared at the end of the call is safe."
```

Any opinions :-)?

Reply:
Although I don't have a problem with us clearly taking a position regarding the problem with EIP-1153, maybe it's better to avoid having the discussion/heat brought over to the compiler. Especially since there were several concerns already raised in the discussion over the years and still the decision to not handle the problem was made.

Reply:
I see no problem with strong wording, but if we do that we have to be absolutely sure it's correct. The spec technically does mention this, even if it basically just says it's not its problem :) So it would be better to say that it's a "limitation" or something. But the new version of the warning side-steps this so it's fine too.

Reply:
I'd say `Transient storage as defined by EIP-1153` though to make it clear it's not about transient storage in general but about how it's defined (implying that it could be defined differently, removing the limitation).

Reply:
Yeah, the most important thing at this point is that we actually raise awareness of this issue, s.t. hopefully people will not introduce bugs due to this.